### PR TITLE
Test harness validator

### PR DESCRIPTION
**Pull request by ashishmax31**
Adds a new validator for ensuring the presence of test-harness image in the addon's metadata file.

### CODE REVIEW

**Comment by sugarraysam** [comment]
I cant use the review feature from Github?? Is this because of the conflicting files?

**Comment by ashishmax31** [comment]
@sugarraysam thats weird, I'm able to see my changes

File: pkg/validators/validate_test_harness.go [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+	if err != nil {
+		return false, "Encountered an error when trying to access the remote image", err
+	}
+	if !imagePresent {
+		return false, "Test harness image doesnt exist", nil
+	}
+	return true, "", nil
+}
+
+func checkImageExists(imageUri *imageparser.Reference) (bool, error) {
```
Comment (sugarraysam) [inline pkg/validators/validate_test_harness.go:33]: As `quay.io` has proven quite flaky recently, do you think it's worth adding a retry mechanism for more reliability?

**Review by sugarraysam — APPROVED** [review]
lgtm, please rebase and follow naming convention.

Also, after picking your `AMxxxx` code, it would be nice to add a wiki page: https://github.com/mt-sre/addon-metadata-operator/wiki/AMxxxx.

View `default_channel` example: https://github.com/mt-sre/addon-metadata-operator/wiki/AM0001

**Comment by ashishmax31** [comment]
@sugarraysam Have a look now. I'll add the wiki after this is merged.

File: pkg/validators/am0005_test_harness.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
+// ValidateAddonLabel validates whether the 'label' field under an addon.yaml follows the format 'api.openshift.com/addon-<id>'
+// TODO - remove this validator once we do field level validation
```
Comment (sugarraysam) [inline pkg/validators/am0005_test_harness.go:2]: nit: remove leftover comments

File: pkg/validators/am0005_test_harness.go [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+func ValidateTestHarness(metabundle utils.MetaBundle) (bool, string, error) {
+	res, err := imageparser.Parse(metabundle.AddonMeta.TestHarness)
+	if err != nil {
+		return false, "Failed to parse testharness url", err
+	}
+	if res.Registry() != "quay.io" {
+		return false, "Testharness image is not in the quay.io registry", nil
+	}
+	imagePresent, err := checkImageExists(res, 5)
```
Comment (sugarraysam) [inline pkg/validators/am0005_test_harness.go:35]: nit: add a var named `retryCount` so we understand the code easily -> `checkImageExists(res, retryCount)`

File: pkg/validators/am0005_test_harness.go [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+	}
+	// Retry on 5xx responses
+	if retryableResponse(resp.StatusCode, retryCount) {
+		return checkImageExists(imageUri, retryCount-1)
+	}
+	return false, nil
+}
+
+func retryableError(err error, currentRetryCount int) bool {
+	if err != nil {
```
Comment (sugarraysam) [inline pkg/validators/am0005_test_harness.go:66]: suggested refactor:

```go
func retryableError(err error) bool {
  urlErr, ok := err.(*url.Error)
  return ok && urlErr.Timeout()
}
```

File: pkg/validators/am0005_test_harness.go [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+	if !imagePresent {
+		return false, "Test harness image doesnt exist", nil
+	}
+	return true, "", nil
+}
+
+func checkImageExists(imageUri *imageparser.Reference, retryCount int) (bool, error) {
+	apiUrl := fmt.Sprintf("%s/%s/manifests/%s", quayRegistryApi, imageUri.ShortName(), imageUri.Tag())
+	resp, err := http.Get(apiUrl)
+	if err != nil {
```
Comment (sugarraysam) [inline pkg/validators/am0005_test_harness.go:48]: Suggest a refactor:
- add `time.Sleep()` -> could get fancy and use `retryCount`
- simplify logic in one line
- `retryableError(err)` has 1 arg

```go
if err != nil && retryCount > 0 && retryableRequestError(err) {
  time.Sleep(2 * time.Second)
  return checkImageExists(imageUri, retryCount-1)
}
```

File: pkg/validators/am0005_test_harness.go [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+		if retryableError(err, retryCount) {
+			return checkImageExists(imageUri, retryCount-1)
+		}
+		return false, err
+	}
+	if resp.StatusCode == 200 {
+		return true, nil
+	}
+	// Retry on 5xx responses
+	if retryableResponse(resp.StatusCode, retryCount) {
```
Comment (sugarraysam) [inline pkg/validators/am0005_test_harness.go:59]: Suggested refactor, get rid of extra func:

```go
if retryCount > 0 && resp.StatusCode >= 500 {
  time.Sleep(2 * time.Second)
  checkImageExists(imageUri, retryCount-1)
}
```

**Review by sugarraysam — COMMENTED** [review]
Left few nits, suggested refactors.

Please make sure to test against `managed-tenants` following the validators SOP before merging.

File: pkg/validators/am0005_test_harness.go (hunk 5 above)
Comment (ashishmax31) [reply pkg/validators/am0005_test_harness.go:48]: If we simplify the logic in one line (`err != nil && retryCount > 0 && retryableRequestError(err) `) We'll have to add check `resp ` is not nil for all the lines below it which expect response to be non nil. Also, we still need to add a
branch which returns the error. 
What about this instead?
```
if err != nil {
	if retryCount > 0 && retryableError(err) {
		time.Sleep(2 * time.Second)
		return checkImageExists(imageUri, retryCount-1)
	}
	return false, err
}
```

**Review by sugarraysam — APPROVED** [review]
lgtm

**Pull request merged**